src/ripple/app/misc/ValidatorKeys.h

```diff
@@ -36,13 +36,14 @@ class Config;
 class ValidatorKeys
 {
 public:
-    PublicKey masterPublicKey;
-    PublicKey publicKey;
-    SecretKey secretKey;
+    std::optional<PublicKey> masterPublicKey;
+    std::optional<PublicKey> publicKey;
+    std::optional<SecretKey> secretKey;
```

Comment:
As soon as I started reviewing this pull request I was stumbling across these `optional`s in RCLConsensus.cpp.  I got really uncomfortable because this new interface suggests that any arbitrary combination of the keys may be missing.  There's a lot of validation that needs to happen to use these safely.

I struggled looking for a solution for several hours and went down a couple of dead ends.  But along the way I found out that either _all_ of these keys are valid or _none_ of them are.  That eventually led to this solution.  Suppose you declare `ValidatorKeys` like this:
```
class ValidatorKeys
{
public:
    // Make a struct to group the keys.  Either all keys are valid or none are.
    struct Keys
    {
        PublicKey masterPublicKey;
        PublicKey publicKey;
        SecretKey secretKey;

        Keys() = delete;
        Keys(
            PublicKey const& masterPublic_,
            PublicKey const& public_,
            SecretKey const& secret_)
        : masterPublicKey(masterPublic_)
        , publicKey(public_)
        , secretKey(secret_)
        { }
    };

    std::optional<Keys> keys;
    NodeID nodeID;
    std::string manifest;
    std::uint32_t sequence = 0;

    ValidatorKeys() = delete;
    ValidatorKeys(Config const& config, beast::Journal j);

    bool
    configInvalid() const
    {
        return configInvalid_;
    }

private:
    bool configInvalid_ = false;  //< Set to true if config was invalid
};
```
Now there's only one `optional`: `keys`.  If `keys` has a value, then I can use any and all of the keys it holds.  Only one place to check to validate using any of them.

I tried to shorten the names of the members of `Keys`.  But I was not successful.  There are three obvious shorter names:

- `masterPublic`
- `public`, and
- `secret`.

Unfortunately for this approach, `public` is a reserved word in C++.  So that won't work.  Oh well.

If you use this suggestion, be sure to use `std::optional::emplace()` to initialize `keys` in the constructor.

Thoughts?

Reply:
thank you Scott, I concur with your approach. I had a similar apprehension about the optionals, but I wasn't able to figure out a better solution. I'll work on this approach.

Why do you emphasize on the use of `std::optional::emplace()` ? Is there a performance benefit from the direct-initialization (apart from perfect forwarding) ?

Reply:
Using `emplace()` is not a huge deal, but it's easy to miss and may provide a minor optimization.

In your `ValidatorKeys` constructor from a `Config` you need to initialize `keys`.  With `emplace()` it looks like this:
```
                keys.emplace(m->masterKey, pk, token->validationSecret);
```
Without `emplace()` it might look like this:
```
                keys = Keys(m->masterKey, pk, token->validationSecret);
```
That results in constructing one more `Keys` object than when `emplace()` is used.  That, in turn, involves copying each key an extra time.  Not a killer, but why do the extra copies if they are not needed?

Reply:
okay 👍 
I read that `emplace` constructs objects "in-place" (Docs: https://en.cppreference.com/w/cpp/utility/optional/emplace). Are we constructing a `Keys` object and moving it into the optional's location? I'm trying to understand how the in-place construction occurs (especially for classes with inheritance)

Reply:
A `std::optional` can be thought of as containing two members:

1. A `bool` that indicates whether the `optional` has a value.
2. A C-style array of `std::byte` that has the proper alignment and adequate size to hold the type contained by the `optional`.

When an `optional` is default constructed the `bool` is set to `false` and the C-style array is left uninitialized.

When an empty optional is assigned to, the copy or move constructor is used to copy or move an already constructed contained type into the C-style array owned by the `optional`.  Once that move or copy is complete the `bool` is set to `true`.  Notice that there are two constructor calls: 1) build the type you want outside of the `optional` 2) construct that same type into the C-style array inside the `optional`.

When you call `emplace()` on an empty optional, then the type being emplaced in the optional does not need to previously exist.  Instead, `emplace()` does a "placement new" of the contained type into the C-style array owned by the `optional`.  We only need to call the constructor for the emplaced type once.

Make sense?  Questions?